File: android/functional-tests/screenshots.py
# This file is part of BBCT for Android.
#
# Copyright 2012 codeguru <codeguru@users.sourceforge.net>
#
# BBCT for Android is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# BBCT for Android is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from com.android.monkeyrunner import MonkeyRunner, MonkeyDevice
from optparse import OptionParser
import util

import os
import sys

defaultPackage = 'bbct.android.common'
defaultDbFileName = 'data/bbct.db'
defaultDelay = 30.0

parser = OptionParser(usage='Usage: %prog [options] <APK file> <list activity>')
parser.add_option('-k', '--package', dest='package', help='Android package name', default=defaultPackage)
parser.add_option('-f', '--dbfile', dest='dbfile', help='Path to local database file', default=defaultDbFileName)
parser.add_option('-d', '--delay', dest='delay', type='float', help='delay before taking a screenshot', default=defaultDelay)

try:
    options, (apkFile, activity) = parser.parse_args()
except ValueError:
    parser.print_help()
    sys.exit(1)

package = options.package
dbFileName = "bbct.db"
localDb = options.dbfile
remoteDb = "/data/data/" + package + "/databases/" + dbFileName

runComponent = package + '/' + activity

# Amount of time to sleep in order to allow the Android emulator to finish
# a task before taking a screenshot. This is necessary for my slow-ass computer
delay = options.delay

print("Connecting to device...")
device = MonkeyRunner.waitForConnection()

print("Installing APK: " + apkFile + "...")
if device.installPackage(apkFile):
    print("Reading baseball card data...")
    cards = util.read_card_data('data/cards.csv')

    for card in cards:
        print card

    print("Starting activity: " + runComponent + "...")
    device.startActivity(component=runComponent)
    MonkeyRunner.sleep(delay)
    util.take_screenshot(device, 'start.png', delay)

    print("Menu...")
    device.press('KEYCODE_MENU', MonkeyDevice.DOWN_AND_UP)
    util.take_screenshot(device, 'menu.png', delay)

    print("Add card...")
    device.press('KEYCODE_DPAD_CENTER', MonkeyDevice.DOWN_AND_UP)
    util.take_screenshot(device, 'add-card-blank.png', delay)

    print("Enter data...")
    util.input_card(device, cards[0])
    util.take_screenshot(device, 'add-card-filled.png', delay)

    print("Removing package: " + package + "...")
    device.removePackage(package)

File: gradle/android/src/androidTest/java/bbct/android/common/functional/test/Screenshots.java
/*
 * This file is part of BBCT for Android.
 *
 * Copyright 2016 codeguru <codeguru@users.sourceforge.net>
 *
 * BBCT for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * BBCT for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package bbct.android.common.functional.test;

import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.support.test.uiautomator.UiObject;
import android.support.test.uiautomator.UiObjectNotFoundException;
import android.support.test.uiautomator.UiSelector;
import android.util.Log;

import bbct.android.common.R;
import bbct.android.common.data.BaseballCard;
import bbct.android.common.test.rule.DataTestRule;
import java.io.File;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static junit.framework.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
public class Screenshots extends UiAutomatorTest {
    @Rule
    public DataTestRule dataTestRule = new DataTestRule();

    private static final String TAG = Screenshots.class.getName();
    private static final long TIMEOUT = 5000;

    private Context context;
    private File screenshotDir;
    private int screenshotCount;
    private BaseballCard card;

    @Before
    public void setUp() throws UiObjectNotFoundException {
        super.setUp();
        screenshotCount = 1;
        context = InstrumentationRegistry.getTargetContext();
        screenshotDir = new File(context.getFilesDir(), "screenshots");

        if (!screenshotDir.exists()) {
            assertTrue("Unable to create directory: " + screenshotDir.getName(), screenshotDir.mkdir());
        }

        Log.d(TAG, "screenshotDir: " + screenshotDir);

        card = dataTestRule.getCard(0);
    }

    @Test
    public void takeScreenshots() throws Throwable {
        UiSelector laterSelector = new UiSelector().text(context.getString(R.string.later));
        UiObject laterButton = device.findObject(laterSelector);
        laterButton.click();

        UiSelector addSelector = new UiSelector().description(context.getString(R.string.add_menu));
        UiObject addMenu = device.findObject(addSelector);
        addMenu.click();
        device.waitForWindowUpdate(context.getPackageName(), TIMEOUT);
        takeScreenshot("NewCard");

        device.pressBack();
        device.waitForWindowUpdate(context.getPackageName(), TIMEOUT);
        takeScreenshot("CardList");

        UiSelector playerNameSelector = new UiSelector().text(card.getPlayerName());
        UiObject listItem = device.findObject(playerNameSelector);
        listItem.click();
        device.waitForWindowUpdate(context.getPackageName(), TIMEOUT);
        device.pressBack();
        takeScreenshot("CardDetails");
    }

    private void takeScreenshot(String description) {
        File screenshotFile
                = new File(screenshotDir,
                           String.format("%02d-%s.png", screenshotCount++, description));
        device.takeScreenshot(screenshotFile);
    }
}
